cmd/preferredimports/preferredimports.go

```diff
@@ -0,0 +1,251 @@
+	isTerminal    = terminal.IsTerminal(int(os.Stdout.Fd()))
+	logPrefix     = ""
+	aliases       map[string]string
+)
+
+type analyzer struct {
+	fset      *token.FileSet // positions are relative to fset
+	ctx       build.Context
+	failed    bool
+	donePaths map[string]interface{}
+	errors    []string
+}
+
+func newAnalyzer() *analyzer {
+	ctx := build.Default
+	ctx.CgoEnabled = true
+
+	a := &analyzer{
+		fset:      token.NewFileSet(),
+		ctx:       ctx,
+		donePaths: make(map[string]interface{}),
+	}
+
+	return a
+}
+
+// collect extracts test metadata from a file.
+func (a *analyzer) collect(dir string) {
+	if _, ok := a.donePaths[dir]; ok {
+		return
+	}
+	a.donePaths[dir] = nil
+
+	// Create the AST by parsing src.
+	fs, err := parser.ParseDir(a.fset, dir, nil, parser.AllErrors|parser.ParseComments)
+
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "ERROR(syntax)", logPrefix, err)
+		a.failed = true
+		return
+	}
+
+	for _, p := range fs {
+		// returns first error, but a.handleError deals with it
+		files := a.filterFiles(p.Files)
+		for _, file := range files {
+			replacements := make(map[string]string)
+			pathToFile := a.fset.File(file.Pos()).Name()
+			for _, imp := range file.Imports {
+				importPath := strings.Replace(imp.Path.Value, "\"", "", -1)
+				importName := ""
+				if imp.Name != nil {
+					importName = imp.Name.Name
+				}
+				if alias, ok := aliases[importPath]; ok {
+					if alias != importName {
+						if !*confirm {
+							_, _ = fmt.Fprintf(os.Stderr, "%sERROR wrong alias for import \"%s\" should be %s in file %s\n", logPrefix, importPath, alias, pathToFile)
+						}
+						a.failed = true
+						replacements[importName] = alias
+						imp.Name.Name = alias
+					}
+				}
+			}
+
+			if len(replacements) > 0 {
+				if *confirm {
+					fmt.Printf("%sReplacing imports with aliases in file %s\n", logPrefix, pathToFile)
+					for key, value := range replacements {
+						renameImportUsages(file, key, value)
+					}
+					ast.SortImports(a.fset, file)
+					var buffer bytes.Buffer
+					if err = format.Node(&buffer, a.fset, file); err != nil {
+						panic(fmt.Sprintf("Error formatting ast node after rewriting import.\n%s\n", err.Error()))
+					}
+
+					fileInfo, err := os.Stat(pathToFile)
+					if err != nil {
+						panic(fmt.Sprintf("Error stat'ing file: %s\n%s\n", pathToFile, err.Error()))
+					}
+
+					err = ioutil.WriteFile(pathToFile, buffer.Bytes(), fileInfo.Mode())
+					if err != nil {
+						panic(fmt.Sprintf("Error writing file: %s\n%s\n", pathToFile, err.Error()))
+					}
+				}
+			}
+		}
+	}
+}
+
+func renameImportUsages(f *ast.File, old, new string) bool {
+	var fixed bool
+
+	// Rename top-level old to new, both unresolved names
+	// (probably defined in another file) and names that resolve
+	// to a declaration we renamed.
+	ast.Inspect(f, func(node ast.Node) bool {
+		if node == nil {
+			return false
+		}
+
+		id, ok := node.(*ast.Ident)
```

Comment:
Can't do a multi-line suggestion: we want to avoid renaming package declarations that happen to have the same name as an import alias (eg: `test/e2e/framework/apps/*.go` are in an `apps` package, and also `import apps "k8s.io/api/apps/v1"`)
```go
		switch id := node.(type) {
		case *ast.Package:
			return false
		case *ast.Ident:
			if id.Name == old {
				id.Name = new
			}
		}
```

Reply:
n/m this didn't quite seem to fix it

Reply:
@spiffxp can you give me a specific example entry in `hack/.import-aliases` that does not work right?

Reply:
`"k8s.io/api/apps/v1": "appsv1"`

Reply:
```go
	// use this to avoid renaming the package declaration, eg:
	//   given: package foo; import foo "bar"; foo.Baz, rename foo->qux
	//   yield: package foo; import qux "bar"; qux.Baz
	var pkg *ast.Ident

	// Rename top-level old to new, both unresolved names
	// (probably defined in another file) and names that resolve
	// to a declaration we renamed.
	ast.Inspect(f, func(node ast.Node) bool {
		if node == nil {
			return false
		}
		switch id := node.(type) {
		case *ast.File:
			pkg = id.Name
		case *ast.Ident:
			if pkg != nil && id == pkg {
				return false
			}
			if id.Name == old {
				id.Name = new
			}
		}
		return true
	})
```

Reply:
Ack applying your fix